### PR TITLE
Params arg no params

### PR DESCRIPTION
**Pull request by slide**
### What does this implement/fix? Explain your changes.

Check the last parameter to see if it is params (and not supplied). By setting the `paramsArray` variable, the binding code does the right thing and passes an empty array.

### Does this close any currently open issues?

#943 

### Checklist

Check all those that are applicable and complete.

-   [X] Make sure to include one or more tests for your change
-   [X] If an enhancement PR, please create docs and at best an example
-   [X] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [X] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

File: src/runtime/methodbinder.cs [hunk 1]
```diff
@@ -415,11 +416,32 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
...
-                        // the managed function accepts it - hopefully :]
-                        ? Runtime.PyTuple_GetSlice(args, arrayStart, pyArgCount)
-                        : Runtime.PyTuple_GetItem(args, paramIndex);
+                    if(arrayStart == paramIndex)
+                    {
+                        // for a params method, we may have a sequence or single/multiple items
+                        // here we look to see if the item at the paramIndex is there or not
+                        // and then if it is a sequence itself.
+                        IntPtr item = Runtime.PyTuple_GetItem(args, paramIndex);
+                        Runtime.PyErr_Clear();
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs:425]: That would just eat all exceptions. Should add proper handling.

File: src/runtime/methodbinder.cs [hunk 2]
```diff
@@ -415,11 +416,32 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
...
+                        if(item != IntPtr.Zero && !Runtime.PyString_Check(item) && Runtime.PySequence_Check(item))
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs:426]: Please, add a test for `void Foo(params object[] args)` called from Python like this: `Foo([], [])`. In this case `args` should become an array of two arrays. It is unclear if this situation is handled.

File: src/runtime/methodbinder.cs (hunk 1 above)
Comment (slide) [reply src/runtime/methodbinder.cs:425]: Ok, I will revisit.

File: src/runtime/methodbinder.cs (hunk 2 above)
Comment (slide) [reply src/runtime/methodbinder.cs:426]: Will do!

File: src/runtime/methodbinder.cs (hunk 1 above)
Comment (slide) [reply src/runtime/methodbinder.cs]: I removed this completely based up updates.

File: src/runtime/methodbinder.cs (hunk 2 above)
Comment (slide) [reply src/runtime/methodbinder.cs]: Added test and fixed up code when it was not passing.

**Comment by slide** [comment]
Close and reopen to see if the build is intermittent.

File: src/runtime/methodbinder.cs [hunk 3]
```diff
@@ -415,11 +416,42 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
...
+                                {
+                                    Runtime.XDecref(item);
+                                }
+                            }
+                        }
+                        else
+                        {
+                            doDecref = true;
+                            op = Runtime.PyTuple_GetSlice(args, arrayStart, pyArgCount);
+                        }
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs:448]: Just a suggestion: maybe move params argument handling to a separate function for clarity

Comment (slide) [reply src/runtime/methodbinder.cs:448]: That's a good idea. Would you like it done as part of this PR, or a separate PR?

**Comment by slide** [comment]
Thanks for reopening, got pulled into a meeting and didn't get a chance to reopen.

**Comment by lostmsu** [comment]
@slide better this one. Then I can squash.

File: src/runtime/methodbinder.cs [hunk 4]
```diff
@@ -397,8 +432,9 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
             {
                 var parameter = pi[paramIndex];
                 bool hasNamedParam = kwargDict.ContainsKey(parameter.Name);
+                bool doDecref = false;
```
Comment (lostmsu) [inline src/runtime/methodbinder.cs:435]: NIT: in python docs this situation is called `isNewReference`

**Comment by lostmsu** [comment]
@slide Travis passed. Need to investigate why is it not being picked up.

**Comment by slide** [comment]
Close and reopen to try and get a Travis build responding.

**Comment by filmor** [comment]
Thank you very much :)

**Pull request merged**